### DISCUSSION TITLE
PDFJS 4.x

### DISCUSSION THREAD

**Discussion by escape-llc** [escape-llc/pdf-component-vue#23] [Ideas]
We are looking at trying to support the latest `4.x` version of `pdfjs` but it did not go well!

The package is generated with top-level await and this is causing problems for our `vite` based build chain.  There are definite issues with consuming their package; the demo application does not even start!  Some webpack references fail to resolve or be awaited on, and they are all `undefined` and it crashes very hard.

Anyone with some expertise here please reply on this thread.  If there are `vite` options or plugins that make the TLA consumption "correct" please advise.

In the meantime we are trying to get repro together to make tickets.

**Comment by MaxAntonsson** [comment]
I've tried out bumping the project to the latest release of `pdfjs` (`4.0.269`), and got around the TLA issue in two different ways.

As explained in this pdfjs-issue: https://github.com/mozilla/pdf.js/issues/17245

- Approach 1:
 Add the plugin: [vite-plugin-top-level-await](https://github.com/Menci/vite-plugin-top-level-await)
 [WIP changes](https://github.com/escape-llc/pdf-component-vue/compare/main...MaxAntonsson:pdf-component-vue:feat-pdfjs4-tla-plugin)
 
 
- Approach 2:
 Specify newer esbuild `target` for dependencies
[WIP changes](https://github.com/escape-llc/pdf-component-vue/compare/main...MaxAntonsson:pdf-component-vue:feat-pdfjs4-es2022)

Also had to drop `.js` specificity on imports as well as adding `.mjs` to the worker url, in order to get things to work in both suggestions.

However I'm now stuck at with an error in one of the unit-tests:
```
src/components/__tests__/PdfComponent.spec.js [ src/components/__tests__/PdfComponent.spec.js ]
TypeError: Cannot destructure property 'AbortException' of 'globalThis.pdfjsLib' as it is undefined.
 ❯ node_modules/pdfjs-dist/web/webpack:/pdf.js/web/pdfjs.js:26:3
```

I'm not that well versed in `vitest` so I'm not sure if the issue lies there or somewhere else.
This also seems like a similar issue with pdfjs v4: https://github.com/mozilla/pdf.js/issues/17228

Any suggestions/insights here? Or maybe possible next steps?

**Comment by escape-llc** [comment]
Looked at the diffs and that appears to be everything we tried previously, albeit with an older build of `pdfjs` 4.  We did not try the TLA plugin method, just the esbuild `target`.  The ticket you referenced seems to point to the `pdf_viewer` modules requiring `async` import.  This async failure inevitably results in something trying to access `globalThis.pdfjsLib` which has not resolved.

We use `pdf_viewer` content for exactly one import for the annotation rendering, which requires that `LinkService` thingie.

In this age of module bundling and `import`, why is anything using `globalThis` to store things?

It seems like v4 is not made to be re-packaged, so maybe we need to go with a `peerDependency` approach and require the consumer of this package to host the `pdfjs` dependencies, instead of the component package.  That would at least "force" all of those modules to load as required; the host can simply include the relevant URLs in their top-level HTML.

**Comment by escape-llc** [comment]
Finally made a breakthrough on this!

The trick is where to "import" and it must be outside the library package, in particular your `main.js` in the application.  This is a safe context for top-level `await` so it all works as expected.

The only other trick is removing all `import` statements in the library that refer to `pdfjs-dist` packages, and the dynamically-imported `Module` must be passed along instead and used at runtime; this was easy to remedy.

The final piece is to use the `provide` and `inject` to pass these `Module` into the rest of the application; in particular the `PdfComponent` picks it up that way.

A new function `usePdfjs` is imported into your `main.js` and that is where the magic happens.

```
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'
import { usePdfjs } from '../lib'

const app = createApp(App)

const bundle = {
	pdfjs: await import(/* @vite-ignore */new URL("pdfjs-dist/build/pdf.min.mjs", import.meta.url)),
	viewer: await import(/* @vite-ignore */new URL("pdfjs-dist/web/pdf_viewer.mjs", import.meta.url)),
	workerUrl: new URL("pdfjs-dist/build/pdf.worker.min.mjs", import.meta.url)
};
usePdfjs(app, bundle)

app.use(router)
app.mount('#app')

```

**Reply by MaxAntonsson** [reply]
This is great 🙌 

I also have a use-case where pdfjs would optimally be loaded via CDN due to size concerns in a UMD-build. This structure would also solve this concern. Thank you for your work 🙏
